drivers/power_domain/power_domain_gpio_monitor.c

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT power_domain_gpio_monitor
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(power_domain_gpio_monitor, CONFIG_POWER_DOMAIN_LOG_LEVEL);
+
+struct pd_gpio_monitor_config {
+	struct gpio_dt_spec power_good_gpio;
+};
+
+struct pd_gpio_monitor_data {
+	struct gpio_callback callback;
+	bool is_powered;
+	const struct device *dev;
+};
+
+struct pd_visitor_context {
+	const struct device *domain;
+	enum pm_device_action action;
+};
+
+static int pd_on_domain_visitor(const struct device *dev, void *context)
+{
+	struct pd_visitor_context *visitor_context = context;
+
+	/* Only run action if the device is on the specified domain */
+	if (!dev->pm || (dev->pm->domain != visitor_context->domain)) {
+		return 0;
+	}
+
+	dev->pm->usage = 0;
+	(void)pm_device_action_run(dev, visitor_context->action);
+	return 0;
+}
+
+static void pd_gpio_monitor_callback(const struct device *port,
+				    struct gpio_callback *cb, gpio_port_pins_t pins)
+{
+	struct pd_gpio_monitor_data *data = CONTAINER_OF(cb, struct pd_gpio_monitor_data, callback);
+	const struct pd_gpio_monitor_config *config = data->dev->config;
+	const struct device *dev = data->dev;
+	struct pd_visitor_context context = {.domain = dev};
+	int rc;
+
+	rc = gpio_pin_get_dt(&config->power_good_gpio);
+	if (rc < 0) {
+		LOG_WRN("Failed to read gpio logic level");
+		return;
+	}
+
+	data->is_powered = rc;
+	if (rc == 0) {
+		context.action = PM_DEVICE_ACTION_SUSPEND;
+		(void)device_supported_foreach(dev, pd_on_domain_visitor, &context);
+		context.action = PM_DEVICE_ACTION_TURN_OFF;
+		(void)device_supported_foreach(dev, pd_on_domain_visitor, &context);
+		return;
+	}
+
+	pm_device_children_action_run(data->dev, PM_DEVICE_ACTION_TURN_ON, NULL);
+}
+
+static int pd_gpio_monitor_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	struct pd_gpio_monitor_data *data = dev->data;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_TURN_ON:
+	case PM_DEVICE_ACTION_TURN_OFF:
+		return -ENOTSUP;
```

Comment:
this can cause a problem, if a device under this domain calls `pm_device_runtime_put()`, the device will first be suspended then if this domain reference's count is 0, the subsystem will try to suspend this domain and return it as result of the `pm_device_runtime_pu()` operation.

It was already discussed here, but this domain is not really a power domain driver but a monitor ... it should probably just return 0 here since there is nothing a driver / subsys / application can do about it.

Reply:
Ohh nevermind, you are not handling  `PM_DEVICE_ACTION_SUSPEND`.